README
###############################################################
# RAWFRAMES 0.91
# Written by Jay Greenberg, CCIE #11021 in May 2003 
# Send comments / bugs to me at jay@greenberg.on.ca
# Visit My homepage at http://www.greenberg.on.ca/
#
# Copyright (C) 2003 Jay Greenberg
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

  see INSTALL for installation instructions

  compiled successfully on the following platforms: 
 	Redhat-7.2
 	Redhat-7.3
 	Redhat-9.0
 	Slackware-7.0



CHANGELOG
0.9 Initial Release
0.91 Added IEEE 802.1q support


ip.h
/*

 RAWFRAMES 
 Written by Jay Greenberg, CCIE #11021 in May 2003
 Send comments / bugs to me at jay@greenberg.on.ca
 Visit My homepage at http://www.greenberg.on.ca/

 Copyright (C) 2003 Jay Greenberg

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
                                                                                                             
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
                                                                                                             
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

 *
 ip.h
 *

*/

 struct iphdr {
        unsigned char ihl:4, version:4;
        unsigned char tos;
        unsigned short int tot_len;
        unsigned short int id;
        unsigned short int frag_off;
        unsigned char ttl;
        unsigned char protocol;
        unsigned short int check;
        unsigned int saddr;
        unsigned int daddr;
 }; /* 20 bytes */



class CIPPacket {

private:

 struct iphdr *ip_packet_header;

public:

 CIPPacket();
 ~CIPPacket();

 char Packet[1500];
 int HeaderLength;;
 int TotalLength;

 void set_version(int version);
 void set_tos(int tos);
 void set_id(int id);
 void set_frag_off(int frag_off);
 void set_ttl(int ttl);
 void set_protocol(int protocol);
 void set_check();
 void set_saddr(int saddr);
 void set_daddr(int daddr);
 void set_ihl(int ihl);
 void set_tot_len(int tot_len);

 void encapsulate(char *payload, int *payload_length); // encapsulates 

private:

 unsigned short csum (unsigned short *buf, int nwords); // calc header checksum
 
};


ip.cpp
/*
                                                                                                              
 RAWFRAMES 
 Written by Jay Greenberg, CCIE #11021 in May 2003
 Send comments / bugs to me at jay@greenberg.on.ca
 Visit My homepage at http://www.greenberg.on.ca/
                                                                                                              
 Copyright (C) 2003 Jay Greenberg
                                                                                                              
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
                                                                                                              
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
                                                                                                              
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
                                                                                                              
 * ip.cpp
 */

#include <string.h>
#include <netinet/in.h>

#include "ip.h"

 CIPPacket::CIPPacket(){
	ip_packet_header = (struct iphdr *) Packet; 
	memset (Packet, 0, sizeof(struct iphdr));	/* zero out the buffer */
 }

 CIPPacket::~CIPPacket(){
	
 }

 void CIPPacket::set_version(int version){
  ip_packet_header->version=version;
 }

 void CIPPacket::set_tos(int tos){
  ip_packet_header->tos=tos; 
 }	 

 void CIPPacket::set_id(int id){
  ip_packet_header->id=htons(id);
 }	 

 void CIPPacket::set_frag_off(int frag_off){
  ip_packet_header->frag_off=htonl(frag_off);
 }	 

 void CIPPacket::set_ttl(int ttl){
  ip_packet_header->ttl=ttl;
 }	 

 void CIPPacket::set_protocol(int protocol){
  ip_packet_header->protocol=protocol;
 }	 

 void CIPPacket::set_check(){
  ip_packet_header->check= csum ((unsigned short *)Packet, ip_packet_header->tot_len >> 1);
 }	 

 void CIPPacket::set_saddr(int saddr){
	 ip_packet_header->saddr=saddr;
 }	 

 void CIPPacket::set_daddr(int daddr){
	 ip_packet_header->daddr=daddr;
 }	 


 unsigned short CIPPacket::csum (unsigned short *buf, int nwords) {
    unsigned long sum;
    for (sum = 0; nwords > 0; nwords--)
        sum += *buf++;
    sum = (sum >> 16) + (sum & 0xffff);
    sum += (sum >> 16);
    return ~sum;
 }



/*
 void CIPPacket::set_options(int options){
	 ip_packet_header._options=options;
 }	 

 void CIPPacket::set_payload(int payload){
	 ip_packet_header._payload=payload;
 }	 
*/

 void CIPPacket::set_ihl(int ihl){
	 ip_packet_header->ihl= (ihl >> 2);
 }	 

 void CIPPacket::set_tot_len(int tot_len){
  ip_packet_header->tot_len=htons(tot_len);
 }	 




void CIPPacket::encapsulate(char *payload, int *payload_length) {

  // requires a *pointer* to the current payload and the length, so that 
  // *bota*h can be updated by this function.

  // Set Lengths
  HeaderLength=sizeof(struct iphdr);
  TotalLength= (HeaderLength + *payload_length);

  set_ihl(HeaderLength);
  set_tot_len(TotalLength);

  // checksum
  ip_packet_header->check= csum ((unsigned short *)Packet, TotalLength >> 1);

  // Encapsulate the payload
  memcpy(Packet+HeaderLength,payload,*payload_length);

  // Set the new payload
  memcpy(payload,Packet,TotalLength);
 
  // Set the new length
  *payload_length = TotalLength;

}






frame.h
/*
                                                                                                              
 RAWFRAMES 
 Written by Jay Greenberg, CCIE #11021 in May 2003
 Send comments / bugs to me at jay@greenberg.on.ca
 Visit My homepage at http://www.greenberg.on.ca/
                                                                                                              
 Copyright (C) 2003 Jay Greenberg
                                                                                                              
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
                                                                                                              
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
                                                                                                              
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
                                                                                                              
 */


/* frame.h			*/
/* Class Def the frame object   */

//# include <rtl.h>

#define PACKET_SIZE 1500
#define MS 1000000              /* one ms in ns */
#define PERIOD (2*MS)           /* send period; recv period is half of that */

namespace args {
 extern int verbosity;
 extern int mac_len;
}

struct dot1qhdr // as an add-on to the ethhdr
{
        unsigned short  h_tci;                  /* Tag Control Info	*/
        unsigned short  h_ptid;                 /* packet type ID field */
} ;


struct LLChdr {

 	char SSAP; //  Source Service Access Point;
 	char DSAP; //  Dest SAP
 	char Control;   //  

};

class CLLC_Header {

	private:
	
		struct LLChdr *llc_header;

	public:

		 CLLC_Header();
 		 ~CLLC_Header();

		char LLC[1500];

		int HeaderLength;;
 		int TotalLength;

 		void set_SSAP(char *SSAP);
 		void set_DSAP(char *DSAP);
 		void set_Control(char *Control);

		void encapsulate(char *payload, int *payload_length); // encapsulates

} ;

struct SNAPhdr {

    unsigned char oui[3];
    unsigned short etype __attribute__ ((packed));

};

class CSNAP_Header {

	private:
	
		struct SNAPhdr *snap_header;

	public:

		 CSNAP_Header();
 		 ~CSNAP_Header();

		char SNAP[1500];

		int HeaderLength;;
 		int TotalLength;

 		void set_oui(char *oui);
 		void set_etype(char *etype);

		void encapsulate(char *payload, int *payload_length); // encapsulates

} ;


class CFrame {

	private:
		char _data[PACKET_SIZE];
		struct ethhdr _ether_header; 
		int _ifindex;     	// interface index
		
	public:
		int  set_dmac(char *dmac);		
		int  set_smac(char *smac);		
		void set_ethertype(char *ethertype, int etherlength, int typelen);
		void set_ifindex(char *iface);		

		int  Transmit(char *packet_buffer, int packet_length, int count);		

		CFrame();  // constructor
		~CFrame(); // de-constructor

};

class CDot1q { 

	private: 
		struct dot1qhdr	*_dot1q_header;

	public:
		CDot1q();
		~CDot1q();

		char DOT1Q[ETH_FRAME_LEN];

                int HeaderLength;;
                int TotalLength;

		void set_ptid(char *ethertype, int etherlength, int typelen); // Protocol Type
		void set_tci(int priority, int cfi, int vid); // Tag Control	

		void encapsulate(char *payload, int *payload_length); 

};

int mac_aton(char *amac, char *nmac);


rawframes.h
/*
                                                                                                              
 RAWFRAMES 
 Written by Jay Greenberg, CCIE #11021 in May 2003
 Send comments / bugs to me at jay@greenberg.on.ca
 Visit My homepage at http://www.greenberg.on.ca/
                                                                                                              
 Copyright (C) 2003 Jay Greenberg
                                                                                                              
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
                                                                                                              
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
                                                                                                              
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
                                                                                                              
 */


/*
 * rawframes.h
 */

int main (int argc, char **argv);
int send_frame (CFrame Frame, int iface);
int readargs (int argc, char **argv);
int checkargs ();

// Argument Variables

namespace args {

 //Ethernet
 static char mac_smac[50];
 static char mac_dmac[50];
 static char mac_etype[10];
 int mac_len; // available extern

 //IPv4
 static int ip_version;
 static int ip_tos;
 static int ip_id;
 static int ip_frag_off;
 static int ip_ttl;
 static int ip_protocol;
 static char ip_saddr[50];
 static char ip_daddr[50];

 //802.1q
 static int dot1q_tag;
 static int dot1q_cfi;
 static int dot1q_priority;
 static int dot1q_vlan;

 //LLC
 static char llc_ssap[10];
 static char llc_dsap[10];
 static char llc_cl[10];

 //SNAP
 static char snap_oui[10];
 static char snap_etype[10];

 //Misc
 static char iface[10];
 static int count;
 int verbosity; // available extern
 
 static char iso_layer2[10];
 static char iso_layer3[10];
 static char iso_layer4[10];

}




rawframes.cpp
/*
                                                                                                              
 RAWFRAMES 
 Written by Jay Greenberg, CCIE #11021 in May 2003
 Send comments / bugs to me at jay@greenberg.on.ca
 Visit My homepage at http://www.greenberg.on.ca/
                                                                                                              
 Copyright (C) 2003 Jay Greenberg
                                                                                                              
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
                                                                                                              
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
                                                                                                              
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
                                                                                                              
 */


/*
 * rawframes.cpp
 */

#include "config.h"

#include <stdlib.h>
#include <string.h>

#include <features.h>    /* for the glibc version number */
#if __GLIBC__ >= 2 && __GLIBC_MINOR >= 1
#include <net/ethernet.h>     /* the L2 protocols */
#else
#include <linux/if_ether.h>   /* the L2 protocols */
#endif

#include <arpa/inet.h>
#include <net/if.h>

#include "frame.h"
#include "ip.h"
#include "rawframes.h"

#include "argtable.h"

int main (int argc, char **argv) {

printf("\n  RawFrames version %s Copyright (C) 2003 Jay Greenberg\n\n  This Program comes with ABSOLUTELY NO WARRANTY\n  This is free software, and you are welcome\n  to redistribute it under certain conditions\n  See http://www.gnu.org/copyleft/gpl.html for details.\n\n  Send comments & bugs to %s\n\n", PACKAGE_VERSION, PACKAGE_BUGREPORT);

  int status;
  char payload[1500];  // The running version of the frame
  int payload_length=0;  // the length of such

  memset (payload, 0, 1500);       /* zero out the buffer */

  
  if (!readargs(argc, argv)) { exit(1); }
  (args::verbosity > 4) && printf("Done Reading CLI Arguments.\n");
  if (!checkargs()) { exit(1); }

  (args::verbosity > 1) && printf("Initializing...\n");
  (args::verbosity > 2) && printf("payload length:  %d\n", payload_length);

  //Layer 3 Encap

  if (strcmp(args::iso_layer3, "ipv4")==0) { 

   if ((strcmp(args::iso_layer2, "802.3")==0)||(strcmp(args::iso_layer2, "LLC")==0)) {
    	  fprintf(stderr, "Frame type %s not compatible with ipv4.  Use ETH_II or LLC_SNAP. \n", 
		args::iso_layer2);
    	  exit(1);
   }

   CIPPacket objIP;

   objIP.set_version(args::ip_version);
   objIP.set_tos(args::ip_tos);
   objIP.set_id(args::ip_version);
   objIP.set_frag_off(args::ip_frag_off);
   objIP.set_ttl(args::ip_ttl);
   objIP.set_protocol(args::ip_protocol);
   objIP.set_saddr(inet_addr (args::ip_saddr));
   objIP.set_daddr(inet_addr (args::ip_daddr));
  
   objIP.set_check();
   objIP.encapsulate(payload, &payload_length);

   (args::verbosity > 1) && printf("Added IP Header...\n");
   (args::verbosity > 2) && printf("payload length:  %d\n", payload_length);

  } else if (strcmp(args::iso_layer3, "ipv6")==0) {
    //todo
  }
   
  if ((strcmp(args::iso_layer2, "LLC")==0) ||(strcmp(args::iso_layer2, "LLC_SNAP")==0)){
  	// If this is an 802.2 (LLC) frame, then let's put LLC headers in...

  	if (strcmp(args::iso_layer2, "LLC_SNAP")==0){
  	// If this is an LLC/SNAP frame, then let's put SNAP headers in...

		CSNAP_Header objSNAP;

		objSNAP.set_oui(args::snap_oui);
		objSNAP.set_etype(args::snap_etype);

   		objSNAP.encapsulate(payload, &payload_length);

	   	(args::verbosity > 1) && printf("Added SNAP Header...\n");
   		(args::verbosity > 2) && printf("payload length:  %d\n", payload_length);

  	}

 	CLLC_Header objLLC;

	objLLC.set_SSAP(args::llc_ssap);
	objLLC.set_DSAP(args::llc_dsap);
	objLLC.set_Control(args::llc_cl);

   	objLLC.encapsulate(payload, &payload_length);

   	(args::verbosity > 1) && printf("Added LLC Header...\n");
   	(args::verbosity > 2) && printf("payload length:  %d\n", payload_length);
  
  }

  // Now that we're down to the frame level, we have to do a type/len determination
  int typelen = strcmp(args::iso_layer2, "ETH_II");  // 0 = Type, !0 = Len 
  char ethertype[7]; strcpy(ethertype, args::mac_etype);

  if (args::dot1q_tag) {

   	(args::verbosity > 4) && printf("Creating Dot1q Tag...\n");
 	CDot1q objDot1q;

	// Tag Control Information
	objDot1q.set_tci(args::dot1q_priority,	//802.1p
			 args::dot1q_cfi,	//Canonical Format ID
			 args::dot1q_vlan);	//VLAN ID

	objDot1q.set_ptid(ethertype, payload_length, typelen); typelen=0;

       	strcpy(ethertype,"0x8100");	//for the ethernet_II type frame (8100=802.1q)

   	objDot1q.encapsulate(payload, &payload_length);

   	(args::verbosity > 1) && printf("Added Dot1q Tag...\n");
   	(args::verbosity > 2) && printf("payload length:  %d\n", payload_length);

 }

 CFrame objFrame;
 objFrame.set_ethertype(ethertype, payload_length, typelen);

 if (!objFrame.set_dmac(args::mac_dmac)){
   	fprintf(stderr, "%s: Invalid Hardware Address\n", args::mac_dmac);
   	exit(1);
 }

 if (!objFrame.set_smac(args::mac_smac)){
  	fprintf(stderr, "%s: Invalid Hardware Address\n", args::mac_smac);
 	exit(1);
 }

 objFrame.set_ifindex(args::iface);

 //Bling Bling
 status = objFrame.Transmit(payload,payload_length,args::count);

 (args::verbosity > 1) && printf("Done Transmitting Frames...\n");



 exit(status);


}

int readargs (int argc, char **argv) {

 arg_rec argtable[] =
      {
      {"-i ", "<ifname>",  arg_str, args::iface, NULL, "\t\t\t Interface to Transmit From"},
      {"-mac_smac ", "<hh:hh:hh:hh:hh:hh>",  arg_str, args::mac_smac, NULL, " Source Mac Address"},
      {"-mac_dmac ", "<hh:hh:hh:hh:hh:hh>",  arg_str, args::mac_dmac, NULL, " Destination Mac Address"},
      {"-iso_layer2 ", "<ETH_II|802.3|LLC|LLC_SNAP>",  arg_str, args::iso_layer2, "ETH_II", "\t Frame Type"},
      {"-mac_etype ", "<0xhhhh>",  arg_str, args::mac_etype, "0x0800", "\t\t 802.3 Type"},
      {"-mac_len ", "<bytes>",  arg_int, &args::mac_len, "0", "\t\t Frame Size "},
      {"-dot1q_tag ", "<0|1>",  arg_int, &args::dot1q_tag, "0", "\t\t Add 802.1q Tag "},
      {"-dot1q_priority ", "<0-7>",  arg_int, &args::dot1q_priority, "0", "\t 802.1q Priority (802.1p)"},
      {"-dot1q_cfi ", "<0|1>",  arg_int, &args::dot1q_cfi, "0", "\t\t 802.1q Canonical Format Identifier "},
      {"-dot1q_vlan ", "<0-4095>",  arg_int, &args::dot1q_vlan, "0", "\t\t 802.1q Vlan ID "},
      {"-llc_ssap ", "<0xhh>",  arg_str, &args::llc_ssap, "0xAA", "\t\t LLC SSAP "},
      {"-llc_dsap ", "<0xhh>",  arg_str, &args::llc_dsap, "0xAA", "\t\t LLC DSAP "},
      {"-llc_cl ", "<0xhh>",  arg_str, &args::llc_cl, "0x03", "\t\t LLC Control "},
      {"-snap_oui ", "<0xhhhhhh>",  arg_str, &args::snap_oui, "0x000000", "\t\t SNAP OUI "},
      {"-snap_etype ", "<0xhhhh>",  arg_str, &args::snap_etype, "0x0800", "\t\t SNAP Type "},
      {"-iso_layer3 ", "<ipv4|none>",  arg_str, args::iso_layer3, "none", "\t Packet Type"},
      {"-ip_version ", "<4|6>",  arg_int, &args::ip_version, "4", "\t\t IP Version"},
      {"-ip_tos ", "<tos>",  arg_int, &args::ip_tos, "0", "\t\t IP Type of Service field / DSCP"},
      {"-ip_id ", "<id>",  arg_int, &args::ip_id, "1", "\t\t\t IP ID"},
      {"-ip_frag_off ", "<offset>",  arg_int, &args::ip_frag_off, "0", "\t IP Fragmentation Offset"},
      {"-ip_ttl ", "<ttl>",  arg_int, &args::ip_ttl, "255", "\t\t IP Time To Live"},
      {"-ip_protocol ", "<protocol>",  arg_int, &args::ip_protocol, "114", "\t IP Protocol Number"},
      {"-ip_saddr ", "<ipaddr>",  arg_str, &args::ip_saddr, "0.0.0.0", "\t\t IP Source Address"},
      {"-ip_saddr ", "<ipaddr>",  arg_str, &args::ip_daddr, "0.0.0.0", "\t\t IP Destination Address"},
      {"-count ", "<number>",  arg_int, &args::count, "1", "\t\t Number of Times to Send the Frame"},
      {"-v ", "<1-5>",  arg_int, &args::verbosity, "1", "\t\t\t Verbose: Use -v 0 to Supress Output (df=1)"}
      };
 
  const size_t narg = (sizeof(argtable)/sizeof(arg_rec));

/*-- process the command line args --*/
  if (argc==1)
     {
     /*-- display program usage and exit. --*/
     printf("Usage: %s %s\n", argv[0], arg_syntax(argtable,narg));
     printf("%s\n",arg_glossary(argtable,narg,"   "));
     return 0;
     }
  else
     {
     /*-- scan command line arguments from argv[] --*/
     char cmdline[200], errmsg[200], errmark[200];
     if (!arg_scanargv(argc,argv,argtable,narg,cmdline,errmsg,errmark))
        {
        /*-- arg error occurred, print error message and exit --*/
        printf("ERROR: %s\n", cmdline);
        printf("       %s %s\n", errmark, errmsg);
        return 0;
        }
      }

 // only get here if the arguments were scanned successfully
 return 1;

}

int checkargs() {
 return 1;
}


frame.cpp
/*
                                                                                                              
 RAWFRAMES 
 Written by Jay Greenberg, CCIE #11021 in May 2003
 Send comments / bugs to me at jay@greenberg.on.ca
 Visit My homepage at http://www.greenberg.on.ca/
                                                                                                              
 Copyright (C) 2003 Jay Greenberg
                                                                                                              
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.
                                                                                                              
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
                                                                                                              
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
                                                                                                              
 */


/* frame.cpp */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>

#include <features.h>    /* for the glibc version number */
#if __GLIBC__ >= 2 && __GLIBC_MINOR >= 1
#include <netpacket/packet.h>
#include <net/ethernet.h>     /* the L2 protocols */
#else
#include <linux/if_packet.h>
#include <linux/if_ether.h>   /* the L2 protocols */
#endif

#include "frame.h"

#include <net/if.h>

CFrame::CFrame() {
	// printf("Creating frame...");
}

CFrame::~CFrame() {
	// printf("Destroying frame...");
}


int CFrame::set_dmac (char *dmac) {
 return mac_aton(dmac, (char*)_ether_header.h_dest);
}

int CFrame::set_smac (char *smac) {
 return mac_aton(smac, (char*)_ether_header.h_source);
}

void CFrame::set_ethertype(char *ethertype, int etherlength, int typelen) {
 (args::verbosity > 4) &&  printf("typelen: %d\n", typelen);
 typelen 
   ? ( _ether_header.h_proto = htons(etherlength) ) // !0
   : ( _ether_header.h_proto = htons(strtoul(ethertype,NULL,16)) ); // 0
}

void CFrame::set_ifindex (char *iface) {
 _ifindex = if_nametoindex(iface);
}

int CFrame::Transmit(char *packet_buffer, int packet_length, int count) {

  char* msg;
  msg = _data;

  int i;
  int msg_size;
  int packet_socket;
  struct sockaddr_ll ll_source_addr;
  struct sockaddr_ll ll_dest_addr;
 
  (args::verbosity > 4) &&  printf("Opening Socket...\n");
  if ( (packet_socket = socket(PF_PACKET, SOCK_RAW, _ether_header.h_proto)) == -1 ) {
    perror("rp_send: main(): socket()");
    exit (1);
  }

  /* source address for bind */
  (args::verbosity > 4) &&  printf("Setting SADDR for BIND...\n");
  ll_source_addr.sll_family = AF_PACKET;
  ll_source_addr.sll_protocol = _ether_header.h_proto;
  ll_source_addr.sll_ifindex = _ifindex;
  ll_source_addr.sll_halen = ETH_ALEN;
  memcpy (ll_source_addr.sll_addr, _ether_header.h_source, ETH_ALEN);

  /* destination address for sendto */
  (args::verbosity > 4) &&  printf("Setting DADDR for SendTo...\n");
  ll_dest_addr.sll_family = AF_PACKET;
  ll_dest_addr.sll_protocol = _ether_header.h_proto;
  ll_dest_addr.sll_ifindex = _ifindex;
  ll_dest_addr.sll_halen = ETH_ALEN;
  memcpy (ll_dest_addr.sll_addr, _ether_header.h_dest, ETH_ALEN);

  (args::verbosity > 4) &&  printf("Binding...\n");
  if (bind (packet_socket,
            (struct sockaddr*)&ll_source_addr,
            sizeof(ll_source_addr)) == -1 ) {
    perror("rp_send: main(): bind()");
    exit (1);
  }

  /* msg */
  msg_size=0;

  // note msg size is probably 0 here, but + msg_size for future...
  (args::verbosity > 4) &&  printf("Building Frame...\n"); 
  memcpy(msg + msg_size, &_ether_header, sizeof(_ether_header)); 
  msg_size += sizeof(_ether_header);

  memcpy(msg + msg_size, packet_buffer, packet_length);
  msg_size += packet_length;

  // We do this here because we need to know the current data size
  /*
  (args::verbosity > 4) &&  printf("Padding Length...\n"); 
  char data[(args::mac_len - msg_size)];
  for (i = 0; i < (args::mac_len - msg_size); i++) {
    if (i >= 1500) { continue; } 
    data[i] = i;
  }

  (args::verbosity > 4) &&  printf("Building Frame Again...\n"); 
  memcpy (msg + msg_size, data, sizeof(data));
  msg_size += sizeof(data);
  */

  for (i = 0; i < count; i++) {

    (args::verbosity > 3) &&  printf("Preparing to send frame %d\n", i+1); 
    if ( sendto(packet_socket, msg, msg_size, 0,
                (struct sockaddr*)&ll_dest_addr,
                sizeof(ll_dest_addr)) == -1 ) {
      perror("rp_send: main(): sendto()");
      exit (1);
    }

    (args::verbosity > 0) &&  printf("Sent Frame %d\n", i+1); 

  } // for i

  (args::verbosity > 4) &&  printf("Closing Socket...\n"); 
  close (packet_socket);

  return 0;

}
// LLC

                                                                                
 CLLC_Header::CLLC_Header(){
        llc_header = (struct LLChdr *) LLC;
        memset (LLC, 0, sizeof(struct LLChdr));       /* zero out the buffer */
 }
                                                                                
 CLLC_Header::~CLLC_Header(){
                                                                                
 }


 void CLLC_Header::set_SSAP(char *SSAP){
  (args::verbosity > 4) &&  printf("Setting SSAP... %s\n", SSAP); 
  llc_header->SSAP=strtoul(SSAP,NULL,16);
 }
                                                                                
 void CLLC_Header::set_DSAP(char *DSAP){
  (args::verbosity > 4) &&  printf("Setting DSAP...%s\n", DSAP); 
  llc_header->DSAP=strtoul(DSAP,NULL,16);
 }
                                                                                
 void CLLC_Header::set_Control(char *Control){
  (args::verbosity > 4) &&  printf("Setting Control...%s\n", Control); 
  llc_header->Control=strtoul(Control,NULL,16);
 }

void CLLC_Header::encapsulate(char *payload, int *payload_length) {
                                                                                
  // requires a *pointer* to the current payload and the length, so that
  // *bota*h can be updated by this function.
                                                                                
  // Set Lengths
  HeaderLength=sizeof(struct LLChdr);
  TotalLength= (HeaderLength + *payload_length);

  (args::verbosity > 4) &&  printf("LLC Header Length... %d\n", HeaderLength);
  (args::verbosity > 4) &&  printf("LLC Total Length... %d\n", TotalLength);
                                                                                
  // Encapsulate the payload
  memcpy(LLC+HeaderLength,payload,*payload_length);

  // Set the new payload
  memcpy(payload,LLC,TotalLength);
                                                                                
  // Set the new length
  *payload_length = TotalLength;
                                                                                
}

// SNAP

                                                                                
 CSNAP_Header::CSNAP_Header(){
        snap_header = (struct SNAPhdr *) SNAP;
        memset (SNAP, 0, sizeof(struct SNAPhdr));       /* zero out the buffer */
 }
                                                                                
 CSNAP_Header::~CSNAP_Header(){
                                                                                
 }


 void CSNAP_Header::set_oui(char *oui){
  (args::verbosity > 4) &&  printf("Setting OUI... %s\n", oui); 
  //memcpy(&snap_header->oui, oui, sizeof(snap_header->oui));
  // snap_header->oui, oui, sizeof(snap_header->oui));
 }
                                                                                
 void CSNAP_Header::set_etype(char *etype){
  (args::verbosity > 4) &&  printf("Setting ETYPE...%s\n", etype); 
  snap_header->etype=htons(strtoul(etype,NULL,16));
 }
                                                                                
void CSNAP_Header::encapsulate(char *payload, int *payload_length) {
                                                                                
  // requires a *pointer* to the current payload and the length, so that
  // *bota*h can be updated by this function.
                                                                                
  // Set Lengths
  HeaderLength=sizeof(struct SNAPhdr);
  TotalLength= (HeaderLength + *payload_length);

  (args::verbosity > 4) &&  printf("SNAP Header Length... %d\n", HeaderLength);
  (args::verbosity > 4) &&  printf("SNAP Total Length... %d\n", TotalLength);
                                                                                
  // Encapsulate the payload
  memcpy(SNAP+HeaderLength,payload,*payload_length);

  // Set the new payload
  memcpy(payload,SNAP,TotalLength);
                                                                                
  // Set the new length
  *payload_length = TotalLength;
                                                                                
}

// .1q

                                                                                
 CDot1q::CDot1q(){
     _dot1q_header = (struct dot1qhdr *) DOT1Q;
     memset (DOT1Q, 0, sizeof(struct dot1qhdr));       /* zero out the buffer */
 }
                                                                                
 CDot1q::~CDot1q(){
                                                                                
 }

 // Protocol Type ID
 void CDot1q::set_ptid(char *ethertype, int etherlength, int typelen) {
  (args::verbosity > 4) &&  printf("dot1q typelen: %d\n", typelen);
  typelen 
    ? ( _dot1q_header->h_ptid = htons(etherlength) ) // !0
    : ( _dot1q_header->h_ptid = htons(strtoul(ethertype,NULL,16)) ); // 0
 }

 // Tag Control Information
 void CDot1q::set_tci(int priority,  int cfi, int vid){
  (args::verbosity > 4) &&  printf("Setting dot1q Priority... %d\n", priority); 
  (args::verbosity > 4) &&  printf("Setting dot1q CFI... %d\n", cfi); 
  (args::verbosity > 4) &&  printf("Setting dot1q VID (Vlan)... %d\n", vid); 
  _dot1q_header->h_tci = htons(vid + (priority << 13 ) + (cfi << 12));
 }
                                                                                
 void CDot1q::encapsulate(char *payload, int *payload_length) {
                                                                                
  // requires a *pointer* to the current payload and the length, so that
  // *bota*h can be updated by this function.
                                                                                
  // Set Lengths
  HeaderLength=sizeof(struct dot1qhdr);
  TotalLength= (HeaderLength + *payload_length);

  (args::verbosity > 4) &&  printf("Dot1q Header Length... %d\n", HeaderLength);
  (args::verbosity > 4) &&  printf("Dot1q Total Length... %d\n", TotalLength);
                                                                                
  // Encapsulate the payload
  memcpy(DOT1Q+HeaderLength,payload,*payload_length);

  // Set the new payload
  memcpy(payload,DOT1Q,TotalLength);
                                                                                
  // Set the new length
  *payload_length = TotalLength;
                                                                                
 }


/*
===============================================================
  mac_aton()
===============================================================
*/
int mac_aton(char *amac, char *nmac)
{
  char c;
  int i;
  unsigned int val;

  i = 0;
  while ((*amac != '\0') && (i < ETH_ALEN)) {
    val = 0;
    c = *amac++;
    if (c >= '0' && c <= '9') {
      val = c - '0';
    }
    else if (c >= 'a' && c <= 'f') {
      val = c - 'a' + 10;
    }
    else if (c >= 'A' && c <= 'F') {
      val = c - 'A' + 10;
    }
    else {
      errno = EINVAL;
      return 0;
    }
    val <<= 4;

    c = *amac;
    if (c >= '0' && c <= '9') {
      val |= c - '0';
    }
    else if (c >= 'a' && c <= 'f') {
      val |= c - 'a' + 10;
    }
    else if (c >= 'A' && c <= 'F') {
      val |= c - 'A' + 10;
    }
    else if (c == ':' || c == '\0') {
      val >>= 4;
    }
    else {
      errno = EINVAL;
      return 0;
    }
    if (c != 0) {
      amac++;
    }
    *nmac++ = val & 0xff;
    i++;

    /* We might get a semicolon here - not required. */
    if (*amac == ':') {
      amac++;
    }
  }
  return 1;
}



